{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' can we write a nicer vcf test set? '"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\" can we write a nicer vcf test set? \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import click\n",
    "import vcfpy\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm\n",
    "from pathos.multiprocessing import _ProcessPool as Pool, ThreadPool\n",
    "from multiprocessing import current_process, Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import numpy as np\n",
    "\n",
    "import vcfpy\n",
    "from hgvs import edit\n",
    "from ncls import NCLS\n",
    "\n",
    "\n",
    "class GenomePosition():\n",
    "    genome_pos_pattern = re.compile(r\"(.+):(\\d+)-(\\d+)\")\n",
    "\n",
    "    def __init__(self, chrom, start, end):\n",
    "        self.chrom = chrom\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "\n",
    "    @classmethod\n",
    "    def from_str(cls, pos_str):\n",
    "        match = cls.genome_pos_pattern.match(pos_str)\n",
    "\n",
    "        if not match:\n",
    "            return None\n",
    "\n",
    "        return cls(match[1], int(match[2]) - 1, int(match[3]))\n",
    "\n",
    "    @classmethod\n",
    "    def from_vcf_record(cls, record):\n",
    "        CHROM = record.CHROM.replace(\"chr\", \"\")\n",
    "\n",
    "        affected_ranges = [\n",
    "            vcf_alt_affected_range(record.REF, alt) for alt in record.ALT\n",
    "        ]\n",
    "        start = record.begin + min(map(lambda r: r.start, affected_ranges),\n",
    "                                   default=0)\n",
    "        end = record.begin + max(map(lambda r: r.stop, affected_ranges),\n",
    "                                 default=1)\n",
    "\n",
    "        return cls(CHROM, start, end)\n",
    "\n",
    "    @classmethod\n",
    "    def from_vcf_record_pos(cls, record):\n",
    "        CHROM = record.CHROM.replace(\"chr\", \"\")\n",
    "        return cls(CHROM, record.begin, record.begin + 1)\n",
    "\n",
    "    @classmethod\n",
    "    def from_gtf_record(cls, record):\n",
    "        return cls(record[0].replace(\"chr\", \"\"),\n",
    "                   int(record[3]) - 1, int(record[4]))\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        return self.chrom == other.chrom and self.start == other.start and self.end == other.end\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"%s:%d-%d\" % (self.chrom, self.start + 1, self.end)\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"%s:%d-%d\" % (self.chrom, self.start + 1, self.end)\n",
    "\n",
    "    def __len__(self):\n",
    "        return self.end - self.start\n",
    "\n",
    "    def __contains__(self, other):\n",
    "        return other.chrom == self.chrom and other.start >= self.start and other.end <= self.end\n",
    "\n",
    "    def __and__(self, other):\n",
    "        if self.chrom != other.chrom:\n",
    "            return None\n",
    "\n",
    "        if other.start >= self.end or self.start >= other.end:\n",
    "            return None\n",
    "\n",
    "        return self.__class__(self.chrom, max(self.start, other.start),\n",
    "                              min(self.end, other.end))\n",
    "\n",
    "    # FIXME: This method may be too overloaded...\n",
    "    def shifted_by(self, start, end=None):\n",
    "        if isinstance(start, range):\n",
    "            start, end = start.start, start.stop\n",
    "\n",
    "        if end is None:\n",
    "            end = start\n",
    "\n",
    "        return self.__class__(self.chrom, self.start + start, self.end + end)\n",
    "\n",
    "    def slice_within(self, other):\n",
    "        if self.chrom != other.chrom:\n",
    "            return None\n",
    "\n",
    "        if self.start < other.start or self.end > other.end:\n",
    "            return None\n",
    "\n",
    "        return slice(self.start - other.start, self.end - other.start)\n",
    "\n",
    "\n",
    "class GenomeIntervalTree():\n",
    "    def __init__(self, predicate, records):\n",
    "        self.predicate = predicate\n",
    "        self.records = []\n",
    "\n",
    "        working_tree_map = {}\n",
    "\n",
    "        idx = 0\n",
    "        for record in records:\n",
    "            genome_pos = predicate(record)\n",
    "\n",
    "            if genome_pos is None:\n",
    "                continue\n",
    "\n",
    "            chrom = genome_pos.chrom\n",
    "\n",
    "            if not chrom in working_tree_map:\n",
    "                # (starts, ends, ids)\n",
    "                working_tree_map[chrom] = ([], [], [])\n",
    "\n",
    "            starts, ends, ids = working_tree_map[chrom]\n",
    "            starts.append(genome_pos.start)\n",
    "            ends.append(genome_pos.end)\n",
    "            ids.append(idx)\n",
    "\n",
    "            self.records.append(record)\n",
    "            idx += 1\n",
    "\n",
    "        tree_map = {}\n",
    "\n",
    "        for chrom, (starts, ends, ids) in working_tree_map.items():\n",
    "            tree_map[chrom] = NCLS(np.array(starts, dtype=np.long),\n",
    "                                   np.array(ends, dtype=np.long),\n",
    "                                   np.array(ids, dtype=np.long))\n",
    "\n",
    "        self.tree_map = tree_map\n",
    "\n",
    "    def _intervals(self, chrom):\n",
    "        return self.tree_map[chrom].intervals()\n",
    "\n",
    "    def _make_query_params(self, genome_pos_list):\n",
    "        starts = np.array([genome_pos.start for genome_pos in genome_pos_list])\n",
    "        ends = np.array([genome_pos.end for genome_pos in genome_pos_list])\n",
    "        ids = np.array(list(range(len(genome_pos_list))))\n",
    "\n",
    "        return (starts, ends, ids)\n",
    "\n",
    "    def _pick_best_record(self, from_ids=None, for_pos=None):\n",
    "        if len(from_ids) < 1:\n",
    "            return None\n",
    "\n",
    "        if len(from_ids) == 1:\n",
    "            return self.records[from_ids[0]]\n",
    "\n",
    "        records = [self.records[record_id] for record_id in from_ids]\n",
    "\n",
    "        scored_records = [(record,\n",
    "                           self._compute_jaccard_index(for_pos,\n",
    "                                                       self.predicate(record)))\n",
    "                          for record in records]\n",
    "        sorted_records = sorted(scored_records,\n",
    "                                key=lambda tup: tup[1],\n",
    "                                reverse=True)\n",
    "\n",
    "        return sorted_records[0][0]\n",
    "\n",
    "    def _compute_jaccard_index(self, pos_a, pos_b):\n",
    "        intersection = pos_a & pos_b\n",
    "\n",
    "        if not intersection:\n",
    "            return 0\n",
    "\n",
    "        # The following is equivalent to |A ∩ B| / |A ∪ B|, but avoids computing\n",
    "        # a union.\n",
    "        # |A ∩ B| / (|A| + |B| - |A ∩ B|)\n",
    "        return len(intersection) / (len(pos_a) + len(pos_b) -\n",
    "                                    len(intersection))\n",
    "\n",
    "    def has_overlap(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return False\n",
    "\n",
    "        return tree.has_overlap(genome_pos.start, genome_pos.end)\n",
    "\n",
    "    def get_first_overlap(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return None\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.first_overlap_both(*qparams)\n",
    "\n",
    "        if len(record_ids) < 1:\n",
    "            return None\n",
    "\n",
    "        return self.records[record_ids[0]]\n",
    "\n",
    "    def get_best_overlap(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return None\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.all_overlaps_both(*qparams)\n",
    "\n",
    "        return self._pick_best_record(from_ids=record_ids, for_pos=genome_pos)\n",
    "\n",
    "    def get_all_overlaps(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return []\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.all_overlaps_both(*qparams)\n",
    "\n",
    "        if any(map(lambda r: r >= len(self.records), record_ids)):\n",
    "            print(\"uh oh i dont like this\")\n",
    "            print(len(self.records))\n",
    "            print(record_ids)\n",
    "\n",
    "        return [self.records[record_id] for record_id in record_ids]\n",
    "\n",
    "    def get_first_containment(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return None\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.all_containments_both(*qparams)\n",
    "\n",
    "        if len(record_ids) < 1:\n",
    "            return None\n",
    "\n",
    "        return self.records[record_ids[0]]\n",
    "\n",
    "    def get_best_containment(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return None\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.all_containments_both(*qparams)\n",
    "\n",
    "        return self._pick_best_record(from_ids=record_ids, for_pos=genome_pos)\n",
    "\n",
    "    def get_all_containments(self, genome_pos):\n",
    "        tree = self.tree_map.get(genome_pos.chrom)\n",
    "\n",
    "        if not tree:\n",
    "            return []\n",
    "\n",
    "        qparams = self._make_query_params([genome_pos])\n",
    "        _, record_ids = tree.all_containments_both(*qparams)\n",
    "\n",
    "        return [self.records[record_id] for record_id in record_ids]\n",
    "\n",
    "\n",
    "class GFFFeature():\n",
    "    @classmethod\n",
    "    def parse_gff_attributes(cls, attr_str):\n",
    "        attr_dict = {}\n",
    "        for key, value in (kv_str.split(' ')\n",
    "                           for kv_str in re.split('; ?', attr_str) if kv_str):\n",
    "            if '\"' in value:\n",
    "                value = value[1:-1]\n",
    "            else:\n",
    "                value = int(value)\n",
    "\n",
    "            attr_dict[key] = value\n",
    "\n",
    "        return attr_dict\n",
    "\n",
    "    @property\n",
    "    def is_forward_stranded(self):\n",
    "        return self.strand == '+'\n",
    "\n",
    "    @property\n",
    "    def is_reverse_stranded(self):\n",
    "        return self.strand == '-'\n",
    "\n",
    "    def __init__(self, record):\n",
    "        self.pos = GenomePosition.from_gtf_record(record)\n",
    "        self.source = record[1]\n",
    "        self.type = record[2]\n",
    "        self.score = None if record[5] == '.' else float(record[5])\n",
    "        self.strand = record[6]\n",
    "        self.phase = None if record[7] == '.' else int(record[7])\n",
    "        self.attributes = self.parse_gff_attributes(record[8])\n",
    "\n",
    "\n",
    "def vcf_alt_affected_range(ref, alt):\n",
    "    # TODO: This method currently only deals with simple substitutions.\n",
    "\n",
    "    if alt.type in [vcfpy.SNV, vcfpy.MNV]:\n",
    "        return range(len(ref))\n",
    "    elif alt.type == vcfpy.INS:\n",
    "        return range(2)\n",
    "    elif alt.type == vcfpy.DEL:\n",
    "        return range(1, len(ref))\n",
    "    elif alt.type == vcfpy.INDEL:\n",
    "        return range(len(ref))\n",
    "\n",
    "    raise NotImplementedError()\n",
    "\n",
    "\n",
    "def _seqs_are_equal(seq_a, seq_b, wildcard=None):\n",
    "    if not len(seq_a) == len(seq_b):\n",
    "        return False\n",
    "\n",
    "    for a, b in zip(seq_a, seq_b):\n",
    "        if a == wildcard or b == wildcard:\n",
    "            continue\n",
    "\n",
    "        if not a == b:\n",
    "            return False\n",
    "\n",
    "    return True\n",
    "\n",
    "\n",
    "# This could be extended for other types of `SequenceVariant`s in the future if\n",
    "# needed.\n",
    "def sequence_variants_are_equivalent(seqvar_a,\n",
    "                                     seqvar_b,\n",
    "                                     strict_uncertain=False,\n",
    "                                     strict_unknown=True,\n",
    "                                     strict_silent=False):\n",
    "    \"\"\"Check if `seqvar_a` and `seqvar_b` are equivalent.\n",
    "    Currently only works correctly for protein-level variants.\n",
    "\n",
    "    Parameters\n",
    "    ---------\n",
    "    strict_uncertain : bool\n",
    "        True if variant (position/edit) uncertainty is factored into\n",
    "        this equivalency check. (default False)\n",
    "    strict_unknown : bool\n",
    "        True if unknown sequence units (e.g. 'X' for amino acids) should\n",
    "        not match known sequence units. (default True)\n",
    "    strict_silent : bool\n",
    "        True if synonymous variants (e.g. 'Arg17=') should not match\n",
    "        otherwise equivalent variants. (default False)\n",
    "    \"\"\"\n",
    "\n",
    "    if not seqvar_a.ac == seqvar_b.ac:\n",
    "        return False\n",
    "\n",
    "    if not seqvar_a.type == seqvar_b.type:\n",
    "        return False\n",
    "\n",
    "    sv_type = seqvar_a.type\n",
    "\n",
    "    if sv_type not in [\"p\"]:\n",
    "        raise NotImplementedError()\n",
    "\n",
    "    posedit_a, posedit_b = seqvar_a.posedit, seqvar_b.posedit\n",
    "\n",
    "    if (posedit_a is None) or (posedit_b is None):\n",
    "        return posedit_a is None and posedit_b is None\n",
    "\n",
    "    if strict_uncertain and not posedit_a.uncertain == posedit_b.uncertain:\n",
    "        return False\n",
    "\n",
    "    pos_a, pos_b = posedit_a.pos, posedit_b.pos\n",
    "\n",
    "    # TODO: Handle positional uncertainty\n",
    "\n",
    "    if not pos_a == pos_b:\n",
    "        return False\n",
    "\n",
    "    edit_a, edit_b = posedit_a.edit, posedit_b.edit\n",
    "\n",
    "    if not type(edit_a) is type(edit_b):\n",
    "        print(type(edit_a), type(edit_b))\n",
    "        return False\n",
    "\n",
    "    _seqs_cmp = lambda a, b: _seqs_are_equal(\n",
    "        a, b, wildcard=(None if strict_unknown else 'X'))\n",
    "\n",
    "    if isinstance(edit_a, (edit.AARefAlt, edit.AAFs, edit.AAExt)):\n",
    "        if (edit_a is None) or (edit_b is None):\n",
    "            return edit_a is None and edit_b is None\n",
    "\n",
    "        if not _seqs_cmp(edit_a.ref, edit_b.ref):\n",
    "            return False\n",
    "\n",
    "        if not _seqs_cmp(edit_a.alt, edit_b.alt):\n",
    "            return False\n",
    "\n",
    "        if strict_silent and (not edit_a.ref) and (not edit_a.alt):\n",
    "            return False\n",
    "    else:\n",
    "        raise NotImplementedError()\n",
    "\n",
    "    if isinstance(edit_a, (edit.AAFs, edit.AAExt)):\n",
    "        if not edit_a.length == edit_b.length:\n",
    "            return False\n",
    "\n",
    "    if isinstance(edit_b, (edit.AAExt)):\n",
    "        if not _seqs_cmp(edit_a.aaterm, edit_b.aaterm):\n",
    "            return False\n",
    "\n",
    "    return True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_sample_vcf(in_stream, out_stream, pos):\n",
    "    \"\"\" most of this cribbed from Rohan -> germline-filter module \"\"\"\n",
    "    \n",
    "    cell_vcf = vcfpy.Reader.from_stream(in_stream)\n",
    "    out_vcf = vcfpy.Writer.from_stream(out_stream, header=cell_vcf.header)\n",
    "    \n",
    "    for record in cell_vcf:\n",
    "        # If a record's ID field is `.`, that means that the calling software\n",
    "        # did not find an ID for it in the associated database, typically dbSNP.\n",
    "        # This is represented as an empty array (`[]`) in VCFPy.\n",
    "        #if record.ID:\n",
    "        #    # This record is in dbSNP; skip it.\n",
    "        #    continue\n",
    "\n",
    "        genome_pos = GenomePosition.from_vcf_record(record)\n",
    "        if str(genome_pos) == pos:\n",
    "            print(genome_pos)\n",
    "        \n",
    "            # write record\n",
    "            out_vcf.write_record(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17:7674241-7674241\n"
     ]
    }
   ],
   "source": [
    "# this is how you call it \n",
    "cell_vcf_path = '../cerebra/tmp/vcf/M20_1001000412.vcf'\n",
    "out_vcf_path = '../cerebra/tmp/artificial_vcf1/test12_tp53_s241f.vcf'\n",
    "pos_str = '17:7674241-7674241'\n",
    "\n",
    "with open(cell_vcf_path, mode='r') as in_file:\n",
    "            with open(out_vcf_path, mode='w') as out_file:\n",
    "                write_sample_vcf(in_file, out_file, pos_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
